src/Functions/custom/New-AzFunctionApp.ps1

```diff
@@ -0,0 +1,364 @@
+            {
+                $appSettings.Add((NewAppSetting -Name 'APPINSIGHTS_INSTRUMENTATIONKEY' -Value $ApplicationInsightsKey))
+            }
+            elseif ($ApplicationInsightsName)
+            {
+                $appInsightsProject = GetApplicationInsightsProject -Name $ApplicationInsightsName
+                if (-not $appInsightsProject)
+                {
+                    $errorMessage = "Failed to get application insights key for project name '$ApplicationInsightsName'. Please make sure the project exist."
+                    $exception = [System.InvalidOperationException]::New($errorMessage)
+                    ThrowTerminatingError -ErrorId "ApplicationInsightsProjectNotFound" `
+                                          -ErrorMessage $errorMessage `
+                                          -ErrorCategory ([System.Management.Automation.ErrorCategory]::InvalidOperation) `
+                                          -Exception $exception
+                }
+
+                $appSettings.Add((NewAppSetting -Name 'APPINSIGHTS_INSTRUMENTATIONKEY' -Value $appInsightsProject.InstrumentationKey))
+            }
+            else
+            {
+                # Create a new ApplicationInsights
+                try 
+                {
+                    $newAppInsightsProject = Az.Functions.internal\New-AzAppInsights -ResourceGroupName $resourceGroupName `
+                                                                                     -ResourceName $Name  `
+                                                                                     -Location $functionAppDef.Location `
+                                                                                     -Kind web `
+                                                                                     -RequestSource "AzurePowerShell" `
+                                                                                     -ErrorAction Stop
+                    if ($newAppInsightsProject)
+                    {
+                        $appSettings.Add((NewAppSetting -Name 'APPINSIGHTS_INSTRUMENTATIONKEY' -Value $newAppInsightsProject.InstrumentationKey))
+                    }                    
+                }
+                catch
+                {
+                    $warningMessage = 'Unable to create the Application Insights for the Function App. Please use the Azure Portal to manually create and configure the Application Insights, if needed.'
+                    Write-Warning $warningMessage
+                }
+            }
+        }       
+
+        # Set app settings and site configuration
+        $siteCofig.AppSetting = $appSettings
+        $functionAppDef.Config = $siteCofig
+        
+        $null = $PSBoundParameters.Add("SiteEnvelope", $functionAppDef)
+        $null = $PSBoundParameters.Add("ErrorAction", "Stop")
+
+        try
+        {
+            $createdFunctionApp = Az.Functions.internal\New-AzFunctionApp @PSBoundParameters
```

Comment:
So, I don't think this is going to work with AsJob or NoWait, because you are passing that parameter through to this cmdlet.

Ideally, for AsJob/NoWait support, you detect that this parameter is supplied, start a job, and recall this cmdlet without the AsJob / nowait (this is the pattern followed here:)

Note that, I thin it is OK not to have AsJob support this time around if that is what you would prefer for this first preview

Reply:
Here is a sample of whatI am talking about in Storage: https://github.com/Azure/azure-powershell/blob/generation/src/Storage/custom/latest-2019-04-30/New-AzStorageAccount.ps1#L256

Reply:
For preview, ww wither need to fix this or remove the parameters

Reply:
Thank you for the example @markcowl. I implemented AsJob, but I am getting a weird deserialization error. I will look into this after I come back from Ignite. I am commenting out the parameters for now.